comp/otelcol/otlp/configcheck/configcheck.go

```diff
@@ -0,0 +1,84 @@
+
+//go:build otlp
+
+// Package configcheck exposes helpers to fetch config.
+package configcheck
+
+import (
+	"strings"
+
+	"github.com/mohae/deepcopy"
+	"go.opentelemetry.io/collector/confmap"
+
+	"github.com/DataDog/datadog-agent/comp/core/config"
+	coreconfig "github.com/DataDog/datadog-agent/pkg/config/setup"
+)
+
+// ReadConfigSection from a config.Component object.
+func ReadConfigSection(cfg config.Reader, section string) *confmap.Conf {
+	// Viper doesn't work well when getting subsections, since it
+	// ignores environment variables and nil-but-present sections.
+	// To work around this, we do the following two steps:
+
+	// Step one works around https://github.com/spf13/viper/issues/819
+	// If we only had the stuff below, the nil sections would be ignored.
+	// We want to take into account nil-but-present sections.
+	//
+	// Furthermore, Viper returns an `interface{}` nil in the case where
+	// `section` is present but empty: e.g. we want to read
+	//	"otlp_config.receiver", but we have
+	//
+	//         otlp_config:
+	//           receiver:
+	//
+	// `GetStringMap` it will fail to cast `interface{}` nil to
+	// `map[string]interface{}` nil; we use `Get` and cast manually.
+	rawVal := cfg.Get(section)
+	stringMap := map[string]interface{}{}
+	if val, ok := rawVal.(map[string]interface{}); ok {
+		// deep copy since `cfg.Get` returns a reference
+		stringMap = deepcopy.Copy(val).(map[string]interface{})
+	}
+
+	// Step two works around https://github.com/spf13/viper/issues/1012
+	// we check every key manually, and if it belongs to the OTLP receiver section,
+	// we set it. We need to do this to account for environment variable values.
+	prefix := section + "."
+	for _, key := range cfg.AllKeysLowercased() {
+		if strings.HasPrefix(key, prefix) && cfg.IsSet(key) {
+			mapKey := strings.ReplaceAll(key[len(prefix):], ".", confmap.KeyDelimiter)
+			// deep copy since `cfg.Get` returns a reference
+			stringMap[mapKey] = deepcopy.Copy(cfg.Get(key))
+		}
+	}
+	return confmap.NewFromStringMap(stringMap)
+}
+
+// IsEnabled checks if OTLP pipeline is enabled in a given config.
+func IsEnabled(cfg config.Reader) bool {
+	return hasSection(cfg, coreconfig.OTLPReceiverSubSectionKey)
+}
+
+// HasLogsSectionEnabled checks if OTLP logs are explicitly enabled in a given config.
+func HasLogsSectionEnabled(cfg config.Reader) bool {
+	return hasSection(cfg, coreconfig.OTLPLogsEnabled) && cfg.GetBool(coreconfig.OTLPLogsEnabled)
+}
+
+func hasSection(cfg config.Reader, section string) bool {
+	// HACK: We want to mark as enabled if the section is present, even if empty, so that we get errors
+	// from unmarshaling/validation done by the Collector code.
+	//
+	// IsSet won't work here: it will return false if the section is present but empty.
+	// To work around this, we check if the receiver key is present in the string map, which does the 'correct' thing.
+	_, ok := ReadConfigSection(cfg, coreconfig.OTLPSection).ToStringMap()[section]
```

Comment:
maybe something to consider for a follow up PR, but it seems that when calling `IsEnabled ` and `HasLogsSectionEnabled ` we parse the `coreconfig.OTLPSection`. Looking at the code, it seems that we are doing a bit of copying and creation of `confmap`

Maybe we couild hoist the result from calling `ReadConfigSection(cfg, coreconfig.OTLPSection)` using a `sync.Once` and avoid parsing the same section multiple times 😄

Reply:
Good point, created a task to look into improving this.